hwilib/devices/bitbox02.py

```diff
@@ -0,0 +1,681 @@
+        """
+        Initializes a new BitBox02 client instance.
+        """
+        super().__init__(path, password=password, expert=expert)
+        if password != "":
+            raise BadArgumentError(
+                "The BitBox02 does not accept a passphrase from the host. Please enable the passphrase option and enter the passphrase on the device during unlock."
+            )
+
+        hid_device = hid.device()
+        hid_device.open_path(path.encode())
+        self.transport = u2fhid.U2FHid(hid_device)
+        self.device_path = path
+
+        # use self.init() to access self.bb02.
+        self.bb02: Optional[bitbox02.BitBox02] = None
+
+        self.noise_config: BitBoxNoiseConfig = CLINoiseConfig()
+
+    def set_noise_config(self, noise_config: BitBoxNoiseConfig) -> None:
+        self.noise_config = noise_config
+
+    def init(self) -> bitbox02.BitBox02:
+        if self.bb02 is not None:
+            return self.bb02
+
+        for device_info in devices.get_any_bitbox02s():
+            if device_info["path"].decode() == self.device_path:
+                bb02 = bitbox02.BitBox02(
+                    transport=self.transport,
+                    device_info=device_info,
+                    noise_config=self.noise_config,
+                )
+                try:
+                    bb02.check_min_version()
+                except FirmwareVersionOutdatedException as exc:
+                    sys.stderr.write("WARNING: {}\n".format(exc))
+                    raise
+            self.bb02 = bb02
+            return bb02
+        raise Exception(
+            "Could not find the hid device info for path {}".format(self.device_path)
+        )
+
+    def close(self) -> None:
+        self.transport.close()
+
+    def get_master_fingerprint_hex(self) -> str:
+        """
+        HWI by default retrieves the fingerprint at m/ by getting the xpub at m/0', which contains the parent fingerprint.
+        The BitBox02 does not support querying arbitrary keypaths, but has an api call return the fingerprint at m/.
+        """
+        bb02 = self.init()
+        if not bb02.device_info()["initialized"]:
+            raise UnavailableActionError("Not initialized")
+        return bb02.root_fingerprint().hex()
+
+    def prompt_pin(self) -> Dict[str, Union[bool, str, int]]:
+        raise UnavailableActionError(
+            "The BitBox02 does not need a PIN sent from the host"
+        )
+
+    def send_pin(self) -> Dict[str, Union[bool, str, int]]:
+        raise UnavailableActionError(
+            "The BitBox02 does not need a PIN sent from the host"
+        )
+
+    def _get_coin(self) -> bitbox02.btc.BTCCoin:
+        if self.is_testnet:
+            return bitbox02.btc.TBTC
+        return bitbox02.btc.BTC
+
+    def _get_xpub(self, keypath: List[int]) -> str:
+        expected_coin = 1 + HARDENED if self.is_testnet else 0 + HARDENED
+        if len(keypath) == 3:
```

Comment:
Good point. I will instead give a more general error message in case the device returns an error; the issue currently is that the BitBox02 does not give very granular error codes, so it is a bit hard to identify what the actual issue was (keypath, unsupported script, etc.?). 

I will remove the keypath validation here and the rest of the plugin.

Reply:
Done.